elrond-wasm/src/storage/mappers/bi_di_mapper.rs

```diff
@@ -0,0 +1,269 @@
+        + TopDecode
+        + NestedEncode
+        + NestedDecode
+        + 'static
+        + Default
+        + PartialEq
+        + ManagedVecItem,
+{
+    fn get_id_key(&self, value: &V) -> StorageKey<SA> {
+        let mut key = self.base_key.clone();
+        key.append_bytes(VALUE_TO_ID_SUFFIX);
+        key.append_item(value);
+        key
+    }
+
+    fn get_value_key(&self, key: &K) -> StorageKey<SA> {
+        let mut value = self.base_key.clone();
+        value.append_bytes(ID_TO_VALUE_SUFFIX);
+        value.append_item(&key);
+        value
+    }
+
+    pub fn get_id(&self, value: &V) -> K {
+        storage_get(self.get_id_key(value).as_ref())
+    }
+
+    fn set_id(&mut self, value: &V, id: &K) {
+        storage_set(self.get_id_key(value).as_ref(), id);
+    }
+
+    pub fn get_value(&self, id: &K) -> V {
+        storage_get(self.get_value_key(id).as_ref())
+    }
+
+    fn set_value(&mut self, id: &K, value: &V) {
+        storage_set(self.get_value_key(id).as_ref(), value);
+    }
+
+    pub fn insert(&mut self, id: K, value: V) -> bool {
+        if self.id_set_mapper.contains(&id) || self.value_set_mapper.contains(&value) {
+            return false;
+        }
+        self.set_id(&value, &id);
+        self.set_value(&id, &value);
+
+        self.id_set_mapper.insert(id);
+        self.value_set_mapper.insert(value);
+        true
+    }
+
+    pub fn remove_by_id(&mut self, id: &K) -> bool {
+        if self.id_set_mapper.swap_remove(id) {
+            let value = self.get_value(id);
+            storage_clear(self.get_id_key(&value).as_ref());
+            storage_clear(self.get_value_key(id).as_ref());
+            self.value_set_mapper.swap_remove(&value);
+            return true;
+        }
+        false
+    }
+    pub fn remove_by_value(&mut self, value: &V) -> bool {
+        if self.value_set_mapper.swap_remove(value) {
+            let id = self.get_id(value);
+            storage_clear(self.get_id_key(value).as_ref());
+            storage_clear(self.get_value_key(&id).as_ref());
+            self.id_set_mapper.swap_remove(&id);
+            return true;
+        }
+        false
+    }
+
+    pub fn remove_all_by_ids<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = K>,
+    {
+        for item in iter {
+            self.remove_by_id(&item);
+        }
+    }
+
+    pub fn remove_all_by_values<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = V>,
+    {
+        for item in iter {
+            self.remove_by_value(&item);
+        }
+    }
+
+    pub fn get_all_values(&self) -> ManagedVec<SA, V> {
+        let mut result = ManagedVec::new();
+        for value in self.value_set_mapper.iter() {
+            result.push(value);
+        }
+        result
+    }
+
+    pub fn get_all_ids(&self) -> ManagedVec<SA, K> {
+        let mut result = ManagedVec::new();
+        for id in self.id_set_mapper.iter() {
+            result.push(id);
+        }
+        result
+    }
```

Comment:
Maybe these should return an iterator of some kind instead of manually pushing them into a ManagedVec.